Add an RPC method showing wallet transactions that are not in the mempool

alamshafil opened the pull request:
This is the PR that implements the RPC method requested in #2721.

As per the issue: either the transaction hash, the transaction hex, or a JSON representation of the transaction can displayed by setting the verbosity parameter.

 - Verbosity value of `0` returns the transaction hash (default)
 - Verbosity value of `1` returns the transaction hex
 - Verbosity value of `2` returns the transaction JSON (same result as `gettransaction`)

I have some questions:

1. What should we name the RPC method? I was thinking either `listnomempooltransactions` or `liststucktransactions`.
2. What should we do about the `include_watchonly` option? Should we keep it?
3. I was not able to create a transaction that had 0 confirmations and not in the mempool. Due to this, I was not able to confirm if the code works; however, in theory the code should work. Furthermore, I was not able to create the tests for this. *(I was only able to create a 0 confirmation TX)*
4. As of right now, any value less than `0` for the verbosity parameter will be seen as `0` and any value greater than `3` will be seen as a `2`. What do you think of this behavior? Should we throw an error?

patricklodder commented:
First off: thanks!

To answer your questions:

1. I like `liststucktransactions`, but open to suggestions
2. Let's keep it, I think we can test it too
3. I'll help. I have some code for this that I can adapt. Give me a day to send you a PR on your repo
4. Since it's really an enumeration of behavior, I'd suggest that throwing an error is better than guessing

patricklodder commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1571,6 +1571,127 @@ UniValue listtransactions(const JSONRPCRequest& request)
+        CWalletTx *const wtx = (*it).second.first;
+        if (wtx->GetDepthInMainChain() <= 0 && !wtx->InMempool())
+        {
+            if (verbosity == 0) {
+                ret.push_back(wtx->GetHash().ToString());
+            } else if (verbosity == 1) {
+                string strHex = EncodeHexTx(static_cast<CTransaction>(*wtx), RPCSerializationFlags());
+                ret.push_back(strHex);
```
Currently these verbosity formats include `watchonly` even if `include_watchonly` is `false` (explicitly or implicitly.)

Perhaps we can add the `IsFromMe()` condition at the top level `if` statement on line 1660:

```cpp
if (wtx->GetDepthInMainChain() <= 0 && !wtx->InMempool() && wtx->IsFromMe(filter)) {
...
```

and can then remove the other evaluations of `wtx->IsFromMe(filter)`, as that will always be true.

This means that we only show our own, outgoing transactions, but since these are the only ones we are capable of doing something about (`abandontransaction` and ideally re-spend the inputs) as for incoming coin we can in nearly any scenario only rebroadcast and hope for the best (fee and dust limits override CPFP), so this would cover the need discussed in #2721 (being aware of any transactions that may go out when upgrading / lowering dust limits) for a first implementation.

We may be able to expand this functionality a bit in a future iteration, to get more fine-grained controls. The must-have for this feature is knowing which of our transactions are stuck (that we created with 1.14.0), so that we can do something about them.

alamshafil replied:
I completely overlooked that. The changes sound good to me. :+1: I believe it's important to have the core feature working as discussed in #2721.

patricklodder commented:
I have PR'd the rpc-test at https://github.com/alamshafil/dogecoin/pull/1

alamshafil commented:
Merged the PR and ran the test on my end. Test is successful for me!

```
$ qa/rpc-tests/liststucktransactions.py 
Initializing test directory /tmp/test83z925ic/574057
Stopping nodes
Cleaning up
Tests successful
```

patricklodder commented:
I'm ACK on this but need 2nd because I now contributed to this

patricklodder commented:
(would be great to rebase to squash commits 1-4, keep 5 separate and remove the merge commit (automagically when rebasing)

alamshafil commented:
I'm not sure if I did the rebase correctly.

chromatic commented on qa/rpc-tests/liststucktransactions.py:
```diff
@@ -0,0 +1,171 @@
+    def setup_network(self, split = False):
+        self.nodes = self.setup_nodes()
+
+        # set the time explicitly
+        self.setmocktime(int(time.time()))
```
Is this redundant with the assignment in the initializer?

I don't feel strongly about this.

patricklodder replied:
I played with the thought of removing the initializer assignment (this also sets it on the nodes, so this one has to stay) but I thought it safer to initialize to a meaningful value there. May be overly defensive though... 

We can call `incrementmocktime(1)` or something but I felt that would be less clear than just explicitly setting the current time (which is on most platforms a second or 2 later than when we did init, because of the time it takes to start up nodes and verify them)

chromatic replied:
That makes sense. I'm fine with this as it is then.

chromatic commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1571,6 +1571,126 @@ UniValue listtransactions(const JSONRPCRequest& request)
+            "\nExamples:\n"
+            "\nList all transactions hashes in the wallet that do not have a mempool entry.\n"
+            + HelpExampleCli("liststucktransactions", "") +
+            "\nList all transactions hexs in the wallet that do not have a mempool entry \n"
```
Should this be `hexs` or `hexes`?

alamshafil replied:
I believe `hexes` is correct.

chromatic commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1571,6 +1571,126 @@ UniValue listtransactions(const JSONRPCRequest& request)
+
+    UniValue ret(UniValue::VARR);
+
+    const CWallet::TxItems & txOrdered = pwalletMain->wtxOrdered;
```
Minor spacing nit; I'd rather this be `const CWallet::TxItems& txOrdered`. That's not a blocker for this PR (if that's the only thing remaining, I'm okay letting it go).

chromatic commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1571,6 +1571,126 @@ UniValue listtransactions(const JSONRPCRequest& request)
+    for (CWallet::TxItems::const_iterator it = txOrdered.begin(); it != txOrdered.end(); ++it)
+    {
+        CWalletTx *const wtx = (*it).second.first;
+        if (wtx->GetDepthInMainChain() <= 0 && !wtx->InMempool() && wtx->IsFromMe(filter))
```
Is this the right order for the conditionals? Because these comparisons short-circuit, we should figure out which of them is the cheapest and/or most likely to skip any unnecessary work.

My guess is `wtx->IsFromMe(filter)` is in the correct place, but I don't know for sure.

patricklodder replied:
Good point. Most of it will be memory operations. We have some bdb caching for the wallet entry (and we lock so it's extremely unlikely to cache-miss) and mempool tx hashes are always in-memory (as the name implies).  

I expect that:
1. `GetDepthInMainChain()` could be more expensive for old transactions than new ones - need to test
2. `InMempool()` is more expensive when mempool is full vs when it's empty, so this is a variable we do not control
3. `IsFromMe()` is arguably more expensive for large wallets than for small ones, but caching should negate that

I'll try to bench this against the reverse order we use here to find out - I have a very dirty 25MB testnet wallet, hope that's enough to see a difference

chromatic replied:
That's plenty fast right now with a healthy number of transactions. If we see an order of magnitude increase in the number of transactions and this code scales linearly (yeah, I know about assumptions), that still seems like acceptable RPC performance. I'm fine with this unmodified then.

chromatic commented:
The rebase looks fine. I have a couple of minor questions.

The only substantive question is this: is "verbosity" the right name for the first parameter? When I read that, I think the smallest value has the least data and the largest value has the most, which is mostly true. In this case, though, a higher verbosity doesn't include the same data as a lower value. It's not additive.

Before I read this code in detail, I expected verbosity level 1 to include the transaction hex *and* the transaction hash, and verbosity level 2 to include both.

Again, I don't feel strongly about this, but it jumped out at me as different from what I expected with the term "verbosity".

chromatic commented:
Sorry about that; I'm not sure how I accidentally closed this. Reopened!

patricklodder commented:
Re: verbosity.

Here's what we (will) have:

1. The only other endpoint that (in the future with 1.21) has 3 verbosity levels is `getblock`. That works as follows:
    - `verbosity=0`: fully serialized block in hex
    - `verbosity=1`: json block with transactions listed only with their hash, including hex
    - `verbosity=2`: json block with fully exploded json-formatted transactions, including hex
2. `getrawmempool` has verbosity with a boolean operator:
    - `verbose=false`: array of transaction ids like this
    - `verbose=true`: full json, including hex

If I were to choose, would propose to remove the int, make it a bool, and do it like `getrawmempool`.

chromatic commented:
The verbosity levels came from the description in #2721, so which approach best fits the requested use case? I too have a preference based on what you describe, but for the purposes I'd use this, it's either verbosity level 2 or verbose true.

alamshafil commented:
I am in favor of using a bool for `verbose`.

>   verbose=false: array of transaction ids like this
>   verbose=true: full json, including hex

If the transaction hex needs to be used, it can be extracted from the full JSON (`verbose=true`).  It can be cheaper to get the TX ID and get the hex through `gettransaction`. 

Unless if having a separate  transaction hex as a output requested in #2721 is necessary, I am in favor of using a boolean.

patricklodder commented:
But yeah I think we agree: let's do the boolean for consistency. The scenario I made for `verbosity=1` above would then become:

```
dogecoin-cli liststucktransactions 1 | jq -r '.[].hex' | dogecoin-tx -json - |jq -r '.vin[] | [.txid, .vout] | join(":")'
```

alamshafil commented:
I have removed the int and replaced it with a bool (`verbose`). I also updated the RPC test to account for the change. I wasn't sure what to do for the transaction hex test, would like feedback on that. (not sure if any() should be used here.)

patricklodder commented:
Re-tested it (on Ubuntu focal) and this works great.

Could use a squash with some reordering... something like this:

```console
$ git rebase -i 1.14.6-dev
pick 7fce9b946 Add RPC method showing wallet TX not in mempool
f    a57d095c3 Convert verbosity to a bool
f    991bf38eb Remove redundant if statement
pick 74269320a qa: add test for liststucktransactions
f    a93135188 qa: Update liststucktransactions test to use verbose parameter
f    e2dd75e79 Use all() instead of any() for test
```

chromatic commented:
ACK, built and tested the full suite here. Looks good to me. Thank you, @alamshafil!

alamshafil commented:
Glad I was able to help out!
